### PR TITLE
truncate long string values

### PR DESCRIPTION
**Pull request by an-tex**

### CODE REVIEW

File: logstage-elastic/src/main/scala/Logging.scala [hunk 1]
```diff
@@ -17,8 +17,15 @@ trait Logging extends LoggerTags {
 
   implicit lazy val log: IzLogger = {
     val legacyNameNormalisation = sys.props.get("logger.izumi.legacyNameNormalisation").contains("true")
-
-    lazy val jsonPolicy = LogstageCirceElasticRenderingPolicy(loggerClass, legacyNameNormalisation)
+    // set this to 0 to turn it off
+    val truncateStringValues = sys.props.get("logger.izumi.truncateStringValuesTo").map(_.toInt)
```
Comment (mariussoutier) [inline logstage-elastic/src/main/scala/Logging.scala:21]: Is it supposed to fail when it cannot be parsed to Long?

File: logstage-elastic/src/main/scala/Logging.scala [hunk 2]
```diff
@@ -17,8 +17,15 @@ trait Logging extends LoggerTags {
...
+    val defaultTruncateStringValuesTo = 16_384 // 16KB
```
Comment (mariussoutier) [inline logstage-elastic/src/main/scala/Logging.scala:22]: or 16 * 1024, but I guess a comment is needed anyway

File: logstage-elastic/src/main/scala/LogstageCirceElasticRenderingPolicy.scala [hunk 3]
```diff
@@ -80,8 +80,17 @@ class LogstageCirceElasticRenderingPolicy(
...
+    dump(maybeTruncatedJson)
+  }
 
-    dump(json)
+  private def truncateStringValues(json: Json, maxLength: Int): Json = {
+    // this is more efficient than json.fold as it avoids unnecessary un- and re-wrapping
+    json
+      .mapObject(_.mapValues(truncateStringValues(_, maxLength)))
+      .mapArray(_.map(truncateStringValues(_, maxLength)))
+      .mapString(string => if (string.length > maxLength) string.take(maxLength) else string)
```
Comment (mariussoutier) [inline logstage-elastic/src/main/scala/LogstageCirceElasticRenderingPolicy.scala:93]: should there be an indicator that the text was shrunk? e.g. `[...]`?

Comment (an-tex) [reply logstage-elastic/src/main/scala/LogstageCirceElasticRenderingPolicy.scala]: good point. done

File: logstage-elastic/src/main/scala/Logging.scala (hunk 1 above)
Comment (an-tex) [reply logstage-elastic/src/main/scala/Logging.scala:21]: yep. I reckon better to fail loud at invalid system properties

File: logstage-elastic/src/main/scala/Logging.scala (hunk 2 above)
Comment (an-tex) [reply logstage-elastic/src/main/scala/Logging.scala]: changed to the multiplication but I left out the bytes suffix. as to be exact it's the length of the string we're referring too (not the length of it's bytes representation which might differ with unicode afaik)

**Comment by mariussoutier** [comment]
Approved

**Pull request merged**